taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -658,6 +658,34 @@ class KernelGen : public IRVisitor {
     emit("}}\n");
   }
 
+  void generate_grid_stride_loop_header() {
+    ScopedIndent _s(line_appender_);
+  }
+
+  struct ScopedGridStrideLoop {
+    KernelGen *gen;
+    std::unique_ptr<ScopedIndent> s;
+
+    ScopedGridStrideLoop(KernelGen *gen) : gen(gen) {
+      size_t stride_size = gen->kernel->program.config.saturating_grid_dim;
+      if (stride_size > 0) {
+        gen->emit("int _sid0 = int(gl_GlobalInvocationID.x) * {};",
+                  stride_size);
+        gen->emit("for (int _sid = _sid0; _sid < _sid0 + {}; _sid++) {{",
```

Comment:
I'm not sure if the stride here is correct - usually the stride is something like `blockDim.x * gridDim.x` instead of `gridDim.x`.

See https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/

Reply:
So, I'm using `gridDim.x` here in cuda term?
Then I'm doing correct, please see
https://github.com/taichi-dev/taichi/blob/master/taichi/backends/opengl/codegen_opengl.cpp#L200-L201

IIUC tpg here is `blockDim.x` in cuda term.

---
stride_size = gridDim
thread = blockDim
work group = somethinglargerDim

Reply:
Thanks for the info. I'm still confused: which part of your code corresponds to `blockDim.x * gridDim.x`?

Reply:
I'm also confused: what does `blockDim` and `gridDim` mean?
IIUC you mean `grid < thread < block`, where `grid`'s are fully serialized for TLS, `thread`'s has shared memory for BLS.
Then OpenGL only provides me `thread < block` (`local thread < work group`). And I have to hand-write a loop for `grid` (`ScoedGridStrideLoop`).

Reply:
Hi, I know that you are confused by my code, but as you can see, CI passed and I didn't observe any performance regression as long as `saturating_grid_dim=0`, given the release of v0.6.21, I'll merge once Travis CI pass to push my 3 draft PR forward, thank for understanding :heart: